spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPushedDownAggregateUtil.java

```diff
@@ -0,0 +1,376 @@
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.ArrayType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.MapType;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import scala.collection.JavaConverters;
+
+/** Helper methods for working with Spark aggregate push down. */
+public class SparkPushedDownAggregateUtil {
+
+  private SparkPushedDownAggregateUtil() {}
+
+  // Build schema for pushed down aggregates. This schema will be used as the scan schema.
+  public static StructType buildSchemaForPushedDownAggregate(
+      List<Expression> aggregates, boolean caseSensitive, Schema schema) {
+    StructType finalSchema = new StructType();
+    for (int index = 0; index < aggregates.size(); index++) {
+      if ((aggregates.get(index)).op().name().equals("COUNTSTAR")) {
+        StructField field =
+            new StructField("COUNT(*)", DataTypes.LongType, false, Metadata.empty());
+        finalSchema = finalSchema.add(field);
+      } else {
+        String colName = ((UnboundAggregate) aggregates.get(index)).ref().name();
+        DataType dataType = getDataTypeForAggregateColumn(colName, caseSensitive, schema);
+        if (dataType instanceof StructType
+            || dataType instanceof ArrayType
+            || dataType instanceof MapType) {
+          // not building pushed down aggregate schema for complex types to disable aggregate push
+          // down because the statistic info for complex are not available.
+          return finalSchema;
+        }
+        if ((aggregates.get(index)).op().name().equals("COUNT")) {
+          StructField field =
+              new StructField(
+                  "COUNT(" + colName + ")", DataTypes.LongType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        } else if ((aggregates.get(index)).op().name().equals("MAX")) {
+          StructField field =
+              new StructField("MAX(" + colName + ")", dataType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        } else if ((aggregates.get(index)).op().name().equals("MIN")) {
+          StructField field =
+              new StructField("MIN(" + colName + ")", dataType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        }
+      }
+    }
+    return finalSchema;
```

Comment:
I think that this should be done in Iceberg code rather than in Spark code. This can produce an Iceberg schema and the convert it. That would allow us to more easily reuse this for other engines.